x-pack/plugins/security/server/authorization/validate_es_response.ts

```diff
@@ -24,35 +24,43 @@ export function validateEsPrivilegeResponse(
   return response;
 }
 
-function buildActionsValidationSchema(actions: string[]) {
-  return schema.object({
-    ...actions.reduce<Record<string, any>>((acc, action) => {
-      return {
-        ...acc,
-        [action]: schema.boolean(),
-      };
-    }, {}),
-  });
-}
-
 function buildValidationSchema(application: string, actions: string[], resources: string[]) {
-  const actionValidationSchema = buildActionsValidationSchema(actions);
+  const actionValidationSchema = schema.boolean();
+  const actionsValidationSchema = schema.object(
+    {},
+    {
+      unknowns: 'allow',
+      validate: (value) => {
+        const actualActions = Object.keys(value).sort();
+        if (
+          actions.length !== actualActions.length ||
+          ![...actions].sort().every((x, i) => x === actualActions[i])
+        ) {
+          throw new Error('Payload did not match expected actions');
+        }
 
-  const resourceValidationSchema = schema.object(
+        Object.values(value).forEach((actionResult) => {
+          actionValidationSchema.validate(actionResult);
+        });
+      },
+    }
+  );
+
+  const resourcesValidationSchema = schema.object(
     {},
     {
       unknowns: 'allow',
       validate: (value) => {
         const actualResources = Object.keys(value).sort();
         if (
           resources.length !== actualResources.length ||
-          !resources.sort().every((x, i) => x === actualResources[i])
+          ![...resources].sort().every((x, i) => x === actualResources[i])
```

Comment:
note: `resources` is provided by consumers, so we shouldn't be mutating their copy of the data.

According to [this benchmark](https://jsbench.me/ofkkpit5k2/1), using the spread operator is the fastest way to copy an array.